# 굿모닝 코딩

https://school.programmers.co.kr/learn/courses/30/lessons/468371


어떤 도로에 차량 신호등이 n개 있습니다.

모든 신호등은 항상 초록불 → 노란불 → 빨간불 순서로 반복되며, 각 신호의 지속 시간은 신호등마다 다릅니다.

시간은 1초부터 시작하며, 각 신호등은 처음에는 초록불 상태로 시작합니다.

이 도로에서는 가끔 정전이 일어나는데,

모든 신호등이 모두 노란불이 되면 정전이 발생한다는 사실이 밝혀졌습니다.

예를 들어 신호등이 2개이고, 각 신호등의 주기가 다음과 같다고 가정해 보겠습니다.

# 논리 순서

1. 입력에 따라 신호등 색깔과 순서 정의
2. G->Y->R 순서로 나열
3. 각각의 숫자만큼 패턴 반복
4. n 개의 신호등 모든 것이 Y 로 나타나는 구간 도출(x초)

In [7]:
# 풀이
signals1 = [[2, 1, 2], [5, 1, 1]]                       # 13
signals2 = [[2, 3, 2], [3, 1, 3], [2, 1, 1]]            # 11
signals3 = [[3, 3, 3], [5, 4, 2], [2, 1, 2]]            # 193
signals4 = [[1, 1, 4], [2, 1, 3], [3, 1, 2], [4, 1, 1]] # -1

def solution(signals):
  time = 1
  MAX_TIME = 1500000                          ## 안겹칠 경우 : 모든 채점 통과를 위한 수치
  while time <= MAX_TIME :                    ## 무한 반복 방지
    all_yellow = True                         # '모든 신호등이 노란불이다' 를 기준으로 함
    for idx, signal in enumerate(signals):
      green = signal[0]
      yellow = signal[1]
      red = signal[2]
      total_cycle = sum(signal)
      remain = time % total_cycle              # 나누기를 통한 나머지를 활용해 현재 위치 확인 // [1/5] 로 나머지가 1이니 첫번째 위치(초록불 상태) 확인
      if remain == 0:                          # 나머지가 0인 경우는 맨 마지막 위치인 것
        remain = total_cycle                     #  [5/5] 는 나머지가 0이지만 마지막 위치(빨간불 상태) 확인
      if green < remain <= (green + yellow):   # 노란불 구간 검사: 초록불 초과 ~ (초록불 + 노란불) 이하인지 확인 //// 3초일때, 6초일때 노란불
        pass                                     # 조건에 맞으면 노란불이므로 아무것도 안 하고 넘어감 (다음 검사 진행)
      else:
        all_yellow = False                     # 신호등이 노란불이 아닌 경우에는 더이상 진행할 필요없음
        break                                    # for 문 탈출하고, time에 +1 함
    if all_yellow :
      return time
    time += 1
  return -1

print(solution(signals1))
print(solution(signals2))
print(solution(signals3))
print(solution(signals4))

13
11
193
-1


# 1. 데이터 추출 실습

In [13]:
# 데이터 셋 추출

from sklearn.datasets import load_iris   # 머신러닝 라이브러리인 싸이킷런(sklearn)의 데이터셋 주머니(datasets)에서
iris = load_iris()                       # 붓꽃(iris) 데이터를 불러오는 함수인 load_iris를 사용한다.

print(iris.DESCR)                        # DESCR : 설명서

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

# 해석

인스턴스 수: 150개(세 가지 클래스 각각 50개)

속성의 수: 4개의 숫자, 예측 속성 및 클래스

속성 정보:

- 세포 길이(cm)

- 세포 폭(cm)

- 꽃잎 길이(cm)

- 꽃잎 너비(cm)

- 수업:

- 아이리스-세토사

- 아이리스 버시컬러

- 아이리스-버지니카



요약 통계:



============== ==== ==== ======= ===== ====================

최소 최대 평균 SD 클래스 상관관계

============== ==== ==== ======= ===== ====================

세포 길이: 4.3 7.9 5.84 0.83 0.7826

세포 폭: 2.0 4.4 3.05 0.43 -0.4194

꽃잎 길이: 1.0 6.9 3.76 1.76 0.9490 (높음!)

꽃잎 너비: 0.1 2.5 1.20 0.76 0.9565 (높음!)

============== ==== ==== ======= ===== ====================



속성 값 누락: 없음

:클래스 분포: 각 3개의 클래스에 대해 33.3%.

:크리에이터: R.A. 피셔

:기부자: 마이클 마샬 (MARshall%PLU@io.arc.nasa.gov )

날짜: 1988년 7월



R.A. 피셔 경이 처음 사용한 유명한 아이리스 데이터베이스입니다.

이 데이터셋은 피셔의 논문에서 나왔습니다.

R에서와 동일하지만 UCI와는 다릅니다

두 개의 잘못된 데이터 포인트가 있는 머신 러닝 저장소.



이것은 아마도 다음에서 찾을 수 있는 가장 잘 알려진 데이터베이스일 것입니다

패턴 인식 문학. 피셔의 논문은 이 분야의 고전입니다

는 오늘날까지도 자주 언급됩니다. (예를 들어, Duda & Hart를 참조하세요.)

데이터 세트에는 각각 50개의 인스턴스로 구성된 3개의 클래스가 포함되어 있으며, 각 클래스는 다음을 의미합니다

홍채 식물의 종류. 한 클래스는 다른 클래스 2와 선형적으로 분리할 수 있습니다

후자는 서로 선형적으로 분리할 수 없습니다.



.. 드롭다운: 참조



- 피셔, R.A. "분류학적 문제에서 여러 측정값의 사용"

연례 우생학, 7부, 2부, 179-188 (1936); 또한 "기여

수학 통계학" (존 와일리, 뉴욕, 1950).

- 두다, R.O., & 하트, P.E. (1973) 패턴 분류 및 장면 분석.

(Q327.D83) 존 와일리 & 선즈. ISBN 0-471-22361-1. 페이지 218 참조.

- 다사라시, B.V. (1980) "이웃을 돌아보며: 새로운 시스템"

부분 노출 인식을 위한 구조 및 분류 규칙

환경" 패턴 분석 및 기계에 관한 IEEE 트랜잭션

인텔리전스, Vol. PAMI-2, No. 1, 67-71.

- 게이츠, G.W. (1972) "감소된 최근접 이웃 규칙". IEEE 트랜잭션

정보 이론에 대하여, 1972년 5월 431-433.

- 참고: 1988년 MLC 소송, 54-64. 치즈맨 등의 오토클라스 II

개념적 클러스터링 시스템은 데이터에서 3개의 클래스를 찾습니다.

In [33]:
# 실습 1-1 무작위 숫자 추출

import numpy as np

nums = []
while len(nums) < 45 :
  tmp = np.random.randint(150)
  if tmp in nums :
    # print("if")
    continue                     # if문 충족 시 다음 작업 진행
  else :
    # print("else")
    nums.append(tmp)
  # print("====")
print(nums)

len(nums)

[17, 90, 107, 16, 36, 53, 123, 122, 142, 95, 137, 77, 140, 41, 63, 144, 105, 76, 6, 22, 70, 94, 98, 115, 29, 49, 125, 99, 20, 35, 10, 81, 14, 104, 119, 62, 5, 51, 120, 55, 108, 43, 2, 72, 11]


45

In [39]:
# 실습 1-2 choice 활용

print(np.random.choice(150, 45))                # 150 개 중 45개 고르기(중복허용 상태)(복원추출)

print(np.random.choice(150, 45, replace=False)) # 중복 비허용(비복원 추출)

[114 146  73  33 101 118 100   6   5  16  36  11  20  15  30  19  37 106
 149  50 130 101  78  72  48  27 113 109 120 143  28  92 129  13  64  65
 120  57   9  84  78 135  78  13  34]
[136  33  29  50 135 137  86  53  45  75  56  62 141   9 103  69  34 118
  51  91  43   7  89   8  54  26  95 146  73 147  35   5  74  31 109 108
  57 101  78 110   4  10  96 116 142]


In [42]:
# 실습 1-3 적용

tstidx = np.random.choice(150, 45, replace=False)

iris.data                                           # 꽃잎, 꽃받침 길이, 너비 데이터가 numpy array로 저장됨

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [43]:
iris.data[tstidx, :]                # tstidx 값을 위치값으로 보고 data 값 추출

array([[6.2, 2.9, 4.3, 1.3],
       [5.2, 3.4, 1.4, 0.2],
       [6.2, 3.4, 5.4, 2.3],
       [4.5, 2.3, 1.3, 0.3],
       [5.7, 2.6, 3.5, 1. ],
       [4.3, 3. , 1.1, 0.1],
       [6.9, 3.1, 5.4, 2.1],
       [6.5, 2.8, 4.6, 1.5],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 2.5, 5.8, 1.8],
       [7.4, 2.8, 6.1, 1.9],
       [4.7, 3.2, 1.6, 0.2],
       [6.7, 3.1, 4.4, 1.4],
       [6.3, 3.3, 4.7, 1.6],
       [4.9, 3.6, 1.4, 0.1],
       [6.3, 2.5, 5. , 1.9],
       [6.4, 3.1, 5.5, 1.8],
       [6.1, 2.8, 4.7, 1.2],
       [6.3, 2.7, 4.9, 1.8],
       [7.1, 3. , 5.9, 2.1],
       [6.9, 3.1, 4.9, 1.5],
       [5.7, 3.8, 1.7, 0.3],
       [5.8, 4. , 1.2, 0.2],
       [4.6, 3.6, 1. , 0.2],
       [5.2, 4.1, 1.5, 0.1],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.9, 3. , 5.1, 1.8],
       [4.6, 3.1, 1.5, 0.2],
       [5.6, 3. , 4.1, 1.3],
       [5.1, 3.5, 1.4, 0.3],
       [6.7, 3.1, 4.7, 1.5],
       [7.3, 2.9, 6.3, 1.8],
       [6.2, 2.2, 4.5, 1.5],
       [5.9, 3

In [44]:
test_data = iris.data[tstidx, :]    # 순서도 섞으면서 랜덤추출
print(test_data.shape)

(45, 4)


In [45]:
test_lbl = iris.target[tstidx]     # 데이터 값에 대한 라벨
print(test_lbl)

[1 0 2 0 1 0 2 1 2 2 2 0 1 1 0 2 2 1 2 2 1 0 0 0 0 2 2 2 0 1 0 1 2 1 1 0 1
 1 2 1 0 0 2 1 2]


In [47]:
# 0,1,2 의 개수 세기

print(np.unique(test_lbl, return_counts=True))         # 0 14개, 1 15개, 2 16개 -> 층화, 계통, 군집 추출 필요   # unique로 중복제거 하고 오름차순 정렬 + return_counts 로 개수 세기

(array([0, 1, 2]), array([14, 15, 16]))


In [48]:
iris.target # 각 클래스 별 50개씩,3번

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

계통 : 0~49까지 15개씩 방식

층화 : 3의 배수마다 추출 방식

In [50]:
# 계통 추출방식 적용해서 추출하기

np.random.choice(50,15, replace=False)         # 인덱스값으로 활용해서 적용

array([39,  7, 12, 35, 40, 36, 49,  3,  5,  9, 29, 16, 21, 20, 18])

In [53]:
iris.data[iris.target == 0][np.random.choice(50,15, replace=False)]                    # target이 0인 data 들 중 15개 뽑기


array([[5. , 3.5, 1.3, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.1, 3.8, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.8, 3.1, 1.6, 0.2],
       [4.6, 3.4, 1.4, 0.3],
       [5.3, 3.7, 1.5, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.4, 3. , 1.3, 0.2],
       [5.5, 3.5, 1.3, 0.2],
       [5. , 3.3, 1.4, 0.2],
       [4.3, 3. , 1.1, 0.1],
       [4.6, 3.6, 1. , 0.2]])

In [62]:
# 앞서 말한거 이미 있음

from sklearn.model_selection import train_test_split
                                                                         # test 값 조정   # 모집단 라벨의 비율을 계산해서 샘플 데이터 비율 적용  # 시드값 고정
x_train, x_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.3,   stratify = iris.target                         ,random_state=7 )

print(x_train.shape, y_test.shape)         # train data 112, test data 38 // 105개가 아닌 이유..? test 값의 기본값이 0.25

(105, 4) (45,)


In [67]:
print(np.unique(y_test, return_counts=True))

(array([0, 1, 2]), array([15, 15, 15]))


In [64]:
x_train

array([[5.4, 3. , 4.5, 1.5],
       [5.4, 3.4, 1.5, 0.4],
       [7.7, 2.8, 6.7, 2. ],
       [5.8, 2.7, 5.1, 1.9],
       [4.4, 3.2, 1.3, 0.2],
       [4.5, 2.3, 1.3, 0.3],
       [7.2, 3.6, 6.1, 2.5],
       [6.3, 2.5, 5. , 1.9],
       [4.6, 3.4, 1.4, 0.3],
       [6.1, 3. , 4.9, 1.8],
       [5.5, 4.2, 1.4, 0.2],
       [5.6, 2.5, 3.9, 1.1],
       [5.8, 2.6, 4. , 1.2],
       [6.3, 3.3, 6. , 2.5],
       [6.5, 3. , 5.5, 1.8],
       [5.9, 3. , 4.2, 1.5],
       [5.1, 3.7, 1.5, 0.4],
       [5.4, 3.9, 1.7, 0.4],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5. , 1.7],
       [6.7, 3. , 5.2, 2.3],
       [5.5, 2.3, 4. , 1.3],
       [5.6, 3. , 4.1, 1.3],
       [6.7, 3.3, 5.7, 2.1],
       [5.6, 3. , 4.5, 1.5],
       [5. , 2.3, 3.3, 1. ],
       [7.7, 3.8, 6.7, 2.2],
       [5.1, 3.8, 1.6, 0.2],
       [7.6, 3. , 6.6, 2.1],
       [4.6, 3.6, 1. , 0.2],
       [4.4, 3. , 1.3, 0.2],
       [6.9, 3.1, 4.9, 1.5],
       [6. , 3. , 4.8, 1.8],
       [6.9, 3.2, 5.7, 2.3],
       [6.3, 2

# 2. 통계분석(수집 - 정제 - 추정 - 검정)


# 1) 수집

자료 수집  및 목적에 맞게 수집

선별작업 필요(정제 과정으로 이어짐)

# 2) 추정

모수를 추정

좋은 추정 결과를 위해서는 양질의 자료 필요

#3) 검정

주장이나 가설이 옳은지 판단

통계적으로 확인하여 가설의 기각과 채택여부 판단

# 2. 확률 추출

1. 단순추출 : 단순무작위로 추출
2. 계통추출 : 일정간격기준으로 추출
3. 층화추출 : 군집내 이질적, 군집간 동질적
4. 집락추출 : 군집내 동질적, 군집간 이질적



# 층화 추출방식

대한민국 고등학생의 학업 성취도를 조사하기 위해 모집단을 [1학년 주머니], [2학년 주머니], [3학년 주머니]로 나눕니다.

그리고 1학년 주머니에서 100명, 2학년에서 100명, 3학년에서 100명을 각각 무작위로 뽑아 총 300명을 조사합니다.


층 내부(동질): 1학년 주머니 안에는 1학년들만 있으므로 내부 구성원의 성격이 비슷합니다.

층 간(이질): 1학년 주머니와 3학년 주머니는 서로 확연히 다른 성격을 가집니다.

# 집락 추출방식

대한민국 고등학생을 조사하려는데 전국을 다 돌아다니기 힘드니, 전국 고등학교 중 [A고등학교], [B고등학교], [C고등학교] 세 군데(덩어리)만 무작위로 뽑습니다.

그리고 선택된 A, B, C 고등학교의 전교생을 전부 조사해 버립니다. 나머지 학교는 조사하지 않습니다.


집락 내부(이질): A고등학교 안에는 1, 2, 3학년, 공부 잘하는 학생, 못하는 학생 등 다양한 사람들이 섞여 있어 모집단의 축소판처럼 복잡합니다.

집락 간(동질): A고등학교나 B고등학교나 대다수 평범한 일반계 고등학교라면 학교끼리의 전반적인 성격은 비슷비슷합니다.

# 3. 기술통계 - 자료형

1. 명목형
2. 순서형
3. 이산형
4. 연속형

# 3. 기술통계 - 자료정리
1. 표
- 도수분포표
- 분할표
- k차원 분할표

2. 그래프
- 파이차트
- 다이어그램
- 막대그래프
- 히스토그램
- 줄기-잎 그림
- 산점도
- 시계열 그림